crates/oxc_parser/src/lexer/number.rs

```diff
@@ -83,72 +82,121 @@ fn parse_binary(s: &str) -> f64 {
+                result <<= 3;
+                result |= n as u64;
+            }
+            result as f64
+        }
     }
+}
 
+#[cold]
+#[inline(never)]
+#[allow(clippy::cast_precision_loss, clippy::cast_lossless)]
+fn parse_octal_slow(s: &str) -> f64 {
     debug_assert!(!s.is_empty());
 
-    let mut result = 0_u64;
+    let mut result = 0_f64;
     for c in s.as_bytes() {
-        result <<= 3;
-        result |= byte_to_value(*c) as u64;
+        let value = f64::from(octal_byte_to_value(*c));
+        result = result.mul_add(8.0, value);
+    }
+
+    result
+}
+
+// ==================================== HEX ====================================
+
+/// - b'0' is 0x30 and b'9' is 0x39.
+/// - b'A' is 0x41 and b'F' is 0x46.
+/// - b'a' is 0x61 and b'f' is 0x66.
+///
+/// So we can convert from a digit to its value with `c & 15`,
+/// and from `A-F` or `a-f` to its value with `(c & 15) + 9`.
+/// We could use `(c & 7) + 9` for `A-F`, but we use `(c & 15) + 9`
+/// so that both branches share the same `c & 15` operation.
+///
+/// This is produces more slightly more assembly than explicitly matching all possibilities,
+/// but only because compiler unrolls the loop.
+///
+/// <https://godbolt.org/z/5fsdv8rGo>
+const fn hex_byte_to_value(c: u8) -> u8 {
+    debug_assert!((c >= b'0' && c <= b'9') || (c >= b'A' && c <= b'F') || (c >= b'a' && c <= b'f'));
+    if c < b'A' {
+        c & 15 // 0-9
+    } else {
+        (c & 15) + 9 // A-F or a-f
     }
-    result as f64
 }
 
 #[allow(clippy::cast_precision_loss, clippy::cast_lossless)]
 fn parse_hex(s: &str) -> f64 {
-    // b'0' is 0x30 and b'9' is 0x39.
-    // b'A' is 0x41 and b'F' is 0x46.
-    // b'a' is 0x61 and b'f' is 0x66.
-    // So we can convert from a digit to its value with `c & 15`,
-    // and from `A-F` or `a-f` to its value with `(c & 15) + 9`.
-    // We could use `(c & 7) + 9` for `A-F`, but we use `(c & 15) + 9`
-    // so that both branches share the same `c & 15` operation.
-    // This is produces more slightly more assembly than explicitly matching all possibilities,
-    // but only because compiler unrolls the loop.
-    // https://godbolt.org/z/5fsdv8rGo
-    const fn byte_to_value(c: u8) -> u8 {
-        debug_assert!(
-            (c >= b'0' && c <= b'9') || (c >= b'A' && c <= b'F') || (c >= b'a' && c <= b'f')
-        );
-        if c < b'A' {
-            c & 15 // 0-9
-        } else {
-            (c & 15) + 9 // A-F or a-f
+    const MAX_FAST_HEX_LEN: usize = 16;
+
+    debug_assert!(!s.is_empty());
+    debug_assert!(!s.starts_with("0x"));
+
+    let without_leading_zeros = s.trim_start_matches('0');
+    match without_leading_zeros.len() {
+        0 => 0.0,
+        1 => return hex_byte_to_value(without_leading_zeros.as_bytes()[0]) as f64,
+        n if n > MAX_FAST_HEX_LEN => parse_hex_slow(without_leading_zeros),
+        _ => {
+            let mut result = 0_u64;
+            let mut has_seen_non_zero = false;
```

Comment:
I'm not clear on the purpose of `has_seen_non_zero` here. Since leading zeros are trimmed off already, unless I'm missing something, `has_seen_non_zero` will always be set to `true` on the first turn of the loop, and `if !has_seen_non_zero` branch will never be taken.

Even if that wasn't the case, I *suspect* (but don't know for sure) that the extra branching on `has_seen_non_zero` is probably more costly than just performing the 2 (very cheap) bitwise ops.

Reply:
it was my attempt to skip leading zeroes; I forgot to clean this up, apologies